File: python/codeforces/two_buttons/two_buttons.py
from collections import deque

def bfs(start, finish):
    queue = deque()
    queue.appendleft((0, start))
    while queue:
        level, val = queue.pop()
        if val == finish:
            return level
        queue.appendleft((level + 1, val - 1))
        queue.appendleft((level + 1, val * 2))

if __name__ == '__main__':
    count = 0
    start, finish = tuple(map(int, input().split()))
    print(bfs(start, finish))

File: cpp/codeforces/two_buttons/two_buttons.cpp
#include <iostream>
#include <queue>
#include <unordered_set>

using namespace std;

long long bfs(long long start, long long finish) {
    long long count = 0;
    queue<pair<int, int>> queue;
    unordered_set<int> visited;
    queue.push(pair<int, int>(0, start));
    while(!queue.empty()) {
        pair<int, int> item = queue.front();
        queue.pop();
        if(visited.find(item.second) != end(visited)){
           continue; 
        }
        visited.insert(item.second);
        if(item.second == finish) {
            return item.first;
        }
        if(visited.find(item.second - 1) == end(visited)){
            queue.push(pair<int, int>(item.first + 1, item.second - 1));
        }
        if(visited.find(item.second * 2) == end(visited)){
            queue.push(pair<int, int>(item.first + 1, item.second * 2));
        }
    }

}

int main() {

    long long first, second;
    scanf("%lld %lld", &first, &second);
    cout << bfs(first, second) << endl;
}
